### PR TITLE
Reduxify DnsStore in Domains

### PR DESCRIPTION
**Pull request by jsnajdr**
Migrate the Flux store for domain DNS records to a Redux one.

- remove the `needsDns` option from the `DomainManagementData` data provider component. The `DomainManagement.Dns` will declare its data needs in the `connect` wrapper and with `<Query*>` components, like the rest of Calypso. It also replaces the `needsDomains` flag with `<QuerySiteDomains>` et al. The `DomainManagementData` should eventually disappear.
- migrate the Flux store to a Redux one by moving the reducer, actions and selectors. That's quite straighforward, as the Flux store was already using a reducer.
- inline the `dns/utils` functions that only the reducer uses into the `reducer` module. Improves bundle splitting
- use the `Query` components and selectors for site domains and domain DNS in ``DomainManagement.Dns`. Requires a bit of refactoring to always render the `Query` components, both with and without placeholder.

**How to test:**
Edit DNS record for a domain, apply a DNS template for services like Office 365 or Google Apps. Verify that nothing is broken.

### CODE REVIEW

File: client/components/data/query-domain-dns/index.jsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ */
+import React from 'react';
+import { useDispatch } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { fetchDns } from 'state/domains/dns/actions';
+
+export default function QueryDomainDns( { domain } ) {
```
Comment (aidvu) [inline client/components/data/query-domain-dns/index.jsx]: Nice. I like this new way.

**Review by aidvu — COMMENTED** [review]
Nice work. Code looks good. Will test later today.

File: client/my-sites/domains/domain-management/dns/index.jsx [hunk 2]
```diff
@@ -109,6 +100,18 @@ class Dns extends React.Component {
 		);
 	}
 
+	render() {
+		const { showPlaceholder, selectedDomainName, selectedSite } = this.props;
+
+		return (
+			<Fragment>
+				<QuerySiteDomains siteId={ selectedSite.ID } />
```
Comment (aidvu) [inline client/my-sites/domains/domain-management/dns/index.jsx]: It's probably a good idea to rewrite this query component the same way we did for the `QueryDomainDns`. Now it fetches domains every time we open the DNS Editor, and it's kind of an overkill. Tho, we'd prolly need a way to force refresh it then, as some paths shouldn't do caching.

An idea for later @Automattic/cobalt.

**Review by aidvu — APPROVED** [review]
LGTM

File: client/my-sites/domains/domain-management/dns/index.jsx (hunk 2 above)
Comment (jsnajdr) [reply client/my-sites/domains/domain-management/dns/index.jsx]: The current behavior is that we query site domains every time any domains screen is opened, and also when switching between different screens (e.g., DNS, nameservers, transfer, ...). The request is _not_ issued only when another one is already in flight (`isRequestingSiteDomains`).

What you probably mean is to never refetch once a domains list has been loaded from server? I.e., introduce the `hasLoadedFromServer` flag?

Comment (aidvu) [reply client/my-sites/domains/domain-management/dns/index.jsx]: > What you probably mean is to never refetch once a domains list has been loaded from server? I.e., introduce the hasLoadedFromServer flag?

That is right. :) Some screens, like the `/domains` should refetch always. DNS one tho, I don't think it's used for anything except `selectedDomain`.

**Pull request merged**